eth/block-validation/api.go

```diff
@@ -228,6 +233,172 @@ func (r *BuilderBlockValidationRequestV2) UnmarshalJSON(data []byte) error {
+		transactionData[i] = tx
+	}
+
+	withdrawalData := make([]*capella.Withdrawal, len(data.Withdrawals))
+	for i, wd := range data.Withdrawals {
+		withdrawalData[i] = &capella.Withdrawal{
+			Index:          capella.WithdrawalIndex(wd.Index),
+			ValidatorIndex: phase0.ValidatorIndex(wd.Validator),
+			Address:        bellatrix.ExecutionAddress(wd.Address),
+			Amount:         phase0.Gwei(wd.Amount),
+		}
+	}
+
+	baseFeePerGas := new(boostTypes.U256Str)
+	err := baseFeePerGas.FromBig(data.BaseFeePerGas)
+	if err != nil {
+		return nil, err
+	}
+
+	return &capella.ExecutionPayload{
+		ParentHash:    [32]byte(data.ParentHash),
+		FeeRecipient:  [20]byte(data.FeeRecipient),
+		StateRoot:     data.StateRoot,
+		ReceiptsRoot:  data.ReceiptsRoot,
+		LogsBloom:     types.BytesToBloom(data.LogsBloom),
+		PrevRandao:    data.Random,
+		BlockNumber:   data.Number,
+		GasLimit:      data.GasLimit,
+		GasUsed:       data.GasUsed,
+		Timestamp:     data.Timestamp,
+		ExtraData:     data.ExtraData,
+		BaseFeePerGas: *baseFeePerGas,
+		BlockHash:     [32]byte(data.BlockHash),
+		Transactions:  transactionData,
+		Withdrawals:   withdrawalData,
+	}, nil
+}
+
+func (api *BlockValidationAPI) BlockAssembler(params *BlockAssemblerRequest) (*capella.ExecutionPayload, error) {
+	log.Info("DEBUG: Entered the BlockAssembler!!")
+	log.Info("BlockAssembler", "tobTxs", len(params.TobTxs.Transactions), "robPayload", params.RobPayload)
+	transactionBytes := make([][]byte, len(params.TobTxs.Transactions))
+	for i, txHexBytes := range params.TobTxs.Transactions {
+		transactionBytes[i] = txHexBytes[:]
+	}
+	txs, err := engine.DecodeTransactions(transactionBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	robBlock, err := engine.ExecutionPayloadV2ToBlock(params.RobPayload.ExecutionPayload)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO - check for gas limits
+	// TODO - support for payouts
+
+	// check if there are any duplicate txs
+	// we can error out if there is a nonce gap
+	// TODO - don't error out, but drop the duplicate tx in the ROB block
+	log.Info("DEBUG: Checking for duplicate txs")
+	seenTxMap := make(map[common.Hash]struct{})
+	for _, tx := range txs {
+		// If we see nonce reuse in TOB then fail
+		if _, ok := seenTxMap[tx.Hash()]; ok {
+			return nil, errors.New("duplicate tx")
+		}
+		seenTxMap[tx.Hash()] = struct{}{}
+	}
+	for _, tx := range robBlock.Transactions() {
+		// if we see nonce re-use in TOB vs ROB then drop txs the txs in ROB
+		if _, ok := seenTxMap[tx.Hash()]; ok {
+			return nil, errors.New("duplicate tx")
+		}
+		seenTxMap[tx.Hash()] = struct{}{}
+	}
+	log.Info("Done Checking for duplicate txs!")
+
+	withdrawals := make(types.Withdrawals, len(params.RobPayload.ExecutionPayload.Withdrawals))
+	log.Info("DEBUG: Building the final set of withdrawals")
+	log.Info("DEBUG: Withdrawals are ", "withdrawals", params.RobPayload.ExecutionPayload.Withdrawals)
+	for i, withdrawal := range params.RobPayload.ExecutionPayload.Withdrawals {
```

Comment:
why is this necessary ? withdrawals should all be the same regardless. why do we have to copy like this

Reply:
its some golang type thing dude. will see if i can make it look cleaner

Reply:
conversions b/w types. there are some opaque types in the code which makes it a bit weird.